text/0135-template-class-object-binding.md

```diff
@@ -0,0 +1,105 @@
+
+- _Vue_: Use the standard [`class`](https://vuejs.org/guide/essentials/class-and-style) HTML attribute as described in this proposal.
+- _Svelte_: Use a custom template directive called [`class:name`](https://svelte.dev/docs/element-directives#class-name), to conditionally apply classes.
+- _Angular_: Use a custom template directive called [`[ngClass]`](https://angular.io/api/common/NgClass), with a similar API as described in this proposal.
+
+It is also worth mentioning that React has no built-in support for conditional style application. However, most React applications use the [classnames](https://www.npmjs.com/package/classnames) NPM package to fill this gap (6.3M weekly downloads).
+
+The `classSet` utility method exposed from the Salesforce internal `lightning/utils` module offers similar capabilities as the public classnames NPM package. This utility function is used by +500 components in Salesforce's internal code base. As this LWC module isn't exposed on the platform, platform components have to re-implement such utility methods for managing complex styles.
+
+As a side note, modern UI frameworks also have built-in capabilities to manipulate element styles. This capability is out of the scope of this RFC and will be the subject of a follow-up proposal.
+
+## Detailed design
+
+This RFC proposes to change the semantics of the template `class` attribute. On top of accepting a string value, the `class` attribute would now accept objects (plain objects and arrays) representing the CSS classes applied to the elements. The LWC engine would then interpret the object to determine which class names should be applied to the element.
+
+The list of class names applied to a given element is defined as follows:
+
+1. If the value is a `string`, apply all the classes in the value.
+2. If the value is an `object`, iterate over its keys and apply the classes with the keys associated with a truthy value.
+3. If the value is an `array`, iterate over its entries and apply this algorithm recursively for each of its entries.
+4. Ignore all other values: `null`, `undefined`, other primitive values, and complex objects.
+
+| Input                                       | Output          |
+| ------------------------------------------- | --------------- |
+| `'foo bar'`                                 | `'foo bar'`     |
+| `{foo: true, bar: false, 'fiz buz': true }` | `foo fiz buz`   |
+| `[false, 'bar', null]`                      | `'bar'`         |
+| `['foo', { bar: true }, ['baz']]`           | `'foo bar baz'` |
+
+**Is this a breaking change?**
+
+In theory, this proposal doesn't introduce a breaking change. However, it introduces a user-land observable change. Even if the `class` attribute only accepts `string` values today, the LWC engine doesn't warn or throw when passing a value of a different type. When passing a non-string value, the LWC engine automatically coerces the value to a string.
```

Comment:
IMO let's create a custom directive, `lwc:classes={className}`, like svelte does.

Reply:
+1 for this suggestion. The `lwc:` directives behave like macros. In this case the `lwc:class` directive can add call the getter on the bound property and add the additional transformations described in the algorithm above.

Reply:
> IMO let's create a custom directive, `lwc:classes={className}`, like svelte does.

As a side note, Svelte's `class:name` directive can be applied to a single class name, not a set of class names. For instance:

```svelte
<button 
  class:active={isActive} 
  class:disabled={isDisabled} 
  class:brand-variant={isBrandVariant}
></button>
``` 

The equivalent in Vue would be:

```vue
<button class={{ 
    active: isActive, 
    disabled: isDisabled, 
    "brand-variant": isBrandVariant
}}></button>
```

It is also worth pointing out that Svelte's syntax doesn't enable expressing dynamic class names.

That said @abdulsattar original question is still valid. Should we use a different directive than the standard `class` attribute for this feature? 

Outside the cognitive overhead of introducing a new directive, the interaction between the standard `class` attribute and the `lwc:class` attribute isn't really obvious. Vue established a precedence with overloading the `class` attribute semantic and developers appear to be happy with this decision. 

Given the fact that this feature will be conditionally enabled using API versioning, it isn't unreasonable for LWC to follow the same approach.